### PR TITLE
OT: Experimental support for hosting 'ot-types' in a Fluid DDS

### PR DESCRIPTION
**Pull request by DLehenbauer**
WIP includes:
- DDS that hosts the 'ot-json1' tree type.
- Proxy-based observer that converts imperative-style changes to tree to ot-json1 ops.
- Bubblebench.

### CODE REVIEW

File: experimental/dds/ot/src/factory.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        attributes: IChannelAttributes): Promise<ISharedOT> {
+        const OT = new SharedOT(id, runtime, attributes);
+        await OT.load(services);
+        return OT;
+    }
+
+    public create(document: IFluidDataStoreRuntime, id: string): ISharedOT {
+        const OT = new SharedOT(id, document, this.attributes);
+        OT.initializeLocal();
+        return OT;
```
Comment (DLehenbauer) [inline experimental/dds/ot/src/factory.ts:52]: nit: 'OT' shouldn't be capitalized (search and replace error)

File: experimental/dds/ot/src/interfaces.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+ * Licensed under the MIT License.
+ */
+
+import { ISharedObject, ISharedObjectEvents } from "@fluidframework/shared-object-base";
+import { Serializable } from "@fluidframework/datastore-definitions";
+
+export interface ISharedOTEvents<T extends Serializable> extends ISharedObjectEvents {
+    (event: "valueChanged", listener: (value: T) => void);
+    (event: "delete", listener: () => void);
+}
```
Comment (DLehenbauer) [inline experimental/dds/ot/src/interfaces.ts:12]: Delete unused events copied from SharedCell.

File: experimental/dds/ot/src/interfaces.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+export interface ISharedOT<T extends Serializable = any> extends ISharedObject<ISharedOTEvents<T>> {
+    remove(path: (number | string)[], value: Serializable): void;
+    move(fromPath: (number | string)[], toPath: (number | string)[]): void;
+    insert(path: (number | string)[], value: Serializable): void;
+    replace(path: (number | string)[], oldVal: Serializable, newVal: Serializable);
+}
```
Comment (DLehenbauer) [inline experimental/dds/ot/src/interfaces.ts:19]: Currently hard-coded to json1 ops.  Will generalize in a follow up PR.

File: experimental/dds/ot/src/ot.ts [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+     * Get a factory for SharedOT to register with the data store.
+     *
+     * @returns a factory that creates and load SharedOT
+     */
+    public static getFactory(): IChannelFactory {
+        return new OTFactory();
+    }
+
+    // eslint-disable-next-line no-null/no-null
+    private root: Doc = null;
```
Comment (DLehenbauer) [inline experimental/dds/ot/src/ot.ts:55]: The json1 'Doc' type === Jsonable.  Plugging in support for IFluidHandle will be an interesting exercise.

File: experimental/dds/ot/src/test/ot.spec.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+                assert.deepEqual(ot2.get(), expected);
+            };
+
+            it("", () => {
+                ot1.insert(["x"], 1);
+                expect({ x: 1 });
+            });
+        });
+    });
+});
```
Comment (DLehenbauer) [inline experimental/dds/ot/src/test/ot.spec.ts:119]: Currently just a few basic tests to sanity check I am using the json1 type correctly.  Needs a lot more (confict tests, reconnection tests, gc tests, stress tests...)

**Review by SamBroner — APPROVED** [review]
This is insane and so cool!

**Review by ksbrar — APPROVED** [review]
😱😱😱

**Review by tanviraumi — APPROVED** [review]

File: experimental/dds/ot/README.md [hunk 6]
```diff
@@ -0,0 +1,3 @@
+# @fluid-experimental/ot
+
+Documentation available at https://fluidframework.com/apis/cell/.
```
[the remark this answers is not in the document]
Comment (DLehenbauer) [reply experimental/dds/ot/README.md]: I missed that, thanks @tylerbutler !

**Pull request merged**